### PR TITLE
The music is coming

### PR DESCRIPTION
**Pull request by lildadou**
I'm happy to submit this pull request that containing these components :
 - a music title parser, able to parse artist, recording title, audio source, codec, bitrate etc
 - a plugin music parser
 - and an input plugin (from SNEP charts) for a crossmatch usage.

I hope these features will be a first step for a better support of music.

### CODE REVIEW

**Comment by stevezau** [comment]
@lildadou looks like you have done alot of work here..

I'm also working on PR https://github.com/Flexget/Flexget/pull/523 for a rest_api and webui..

I've found it valuable to also join the flexget IRC channel to discuss things in more detail. Just a suggestion.

**Comment by stevezau** [comment]
@gazpachoking or @paranoidi  comments here??

**Comment by lildadou** [comment]
There is no urgency, music plugins are not mature enough to be actually used. Only the parser is truly finished. I am achieving (2-3 weeks) an equivalent of series plugin but for music charts. 
At this moment, I will promote more (and come on IRC) ;)

**Comment by stevezau** [comment]
@lildadou is this still a WIP or is it functional? I notice you joined the IRC channel but that was 3am my time :)

**Comment by lildadou** [comment]
@gazpachoking or @paranoidi : plugins are now usefull and a merging will be valuable for users (especially french users because the only one charts connector is for France).

I am working for audio quality filtering. Quality class is currently limited for music purpose. 
Can we discuss the design about Quality (quality, quality requirements etc)? 

In my opinion, one quality should be used per *stream*. A stream could be a video, audio, image etc.
Next, theses qualities are packaged on a *ReleaseQuality*.
release quality design - new page
Edit this class diagram on [LucidChart](https://www.lucidchart.com/invitations/accept/dd7f276a-cad0-4530-bf18-7a241eceecb8)

**Comment by lildadou** [comment]
*Le texte est disponible en Français plus bas*

## English version
Hello everyone, it seems that things have not progress much.
Here is some documentation to help you review my code.

The project can be divided into several functional units:
  - Parser of music titles
  - The metainfo plugin music  
  - The filter plugins charts
  - An implementation of ChartsConnector, SnepChartsConnector, intended for French users

#### The musical title parser
This parser uses Guessit. Guessit is a parser designed to extract all the information contained in the file name of movie or series episode. To achieve this, Guessit has a modular and sequential approach. Each module is invoked to extract a specific information (the year the video codec, video source etc.). The file name is reduced to every information extracted and the likelihood only remains the title of the film/series episode increases.

The musical title parser works the same way. However, I added more modules to meet the requirements of audio file names. These modules are into flexget/plugins/parsers/music/guess_ *. The parser itself is into the file flexget/plugins/parsers/music/parser_guessit_music.py.
Parsed informations are represented by Plain Old Python Object. You will find their abstract definitions in flexget/plugins/parsers/parser_common.py and an implementation in flexget/plugins/parsers/music/parser_guessit_music.py.

#### The plugin metainfo music
This plugin is a good illustration of using the parser. This plugin adds the title properties, artist, year and quality that have been extracted from the input name.
To achieve this, it get the current music parser, parse the input name with the parser and finally, adds information to the entry using the parser's mapping dictionary.

#### The charts filter plugin
Charts filter musical entries by their charts position and in-charts duration. An entry is accepted if it fulfilled all the conditions of the filter.

Charts is interoperable with any charts provider as long as a connector exists. Such a connector must implement the ChartsConnector interface and be registered as a plugin of charts_connector group.

The ChartsConnector must always provide charts with an expiration date. With this, the charts may retain and re-use local charts retrieved from the Internet. The use of persistence is motivated by the fact that the charts are released than once a week. Persistence is done with 3 tables:
 - ChartsProvider represents a data source. For example, daily charts for US radio provided by WorldCharts.example.com is a ChartsProvider. WorldCharts.example.com alone is not a ChartsProvider.
 - ChartsRelease represents the entries list of charts from a ChartsProvider. A ChartsRelease have an expiration date.
 - ChartsEntry represents a chart's entry. It contains the identifiers of the recording (title, artist) and its position in the charts. Also I included his best charts ranking since its release and the number of week on charts. I don't know if we should leave these two latest information.

#### ChartsConnector for Frenchies
SnepChartsConnector is a ChartsConnector will retrieve the official charts in France on snepmusique.com via a simple HTML extraction.





## French version
Bonjour à tous, il semble que les choses n'ont pas beaucoup avancées.
Voici un peu de documentation pour vous aider à reviewer mon code.

Le projet peut se décomposer en plusieurs unités fonctionnelles :
 - le parser de titres musicaux
 - Le plugin music de type metainfo
 - le plugins charts de type filtre
 - L'implémentation de ChartsConnector, SnepChartsConnector, à destination des usagers français
 
#### Le parser de titre musicaux
Ce parser utilise Guessit. Guessit est parser initialement conçue pour extraire toutes les informations contenue dans le nom de fichier d'un film ou d'un épisode de série. Pour y parvenir, Guessit propose une approche modulaire et séquentielle. Chaque module est invoqué pour extraire une information bien précise (l'année, le codec vidéo, la source vidéo etc). A chaque information extraite, le nom de fichier est réduit et la probabilité qu'il ne reste que le titre de film/épisode de série augmente.

Le parser de titre musicaux fonctionne de la même façon. J'ai réalisé plusieurs modules pour satisfaire les exigences des noms de fichier audios. Ces modules sont flexget/plugins/parsers/music/guess_*. Le parser en lui même est dans le fichier flexget/plugins/parsers/music/parser_guessit_music.py.
Les informations parsées sont représentées par des Plain Old Python Object. Vous trouverez leurs définitions abstraites dans flexget/plugins/parsers/parser_common.py et une implémentation dans flexget/plugins/parsers/music/parser_guessit_music.py.

#### Le plugin metainfo music
Ce plugin est une bonne illustration de l'utilisation du parser. Ce plugin ajoute les propriétés titre, artiste, année et qualité qui ont été extrait du nom de l'entrée.
Pour y parvenir, il récupère simplement le parser de type music, parse le nom de l'entrée grâce au parser et pour terminer, ajoute les informations à l'entrée à l'aide du dictionnaire de mappage fourni par le parser.

#### Le plugin filtre charts
Charts est un plugin qui filtre les entrées musicales en fonction de position du titre pour un charts donné. Une entrée est acceptée si elle remplie toutes les conditions du filtre.

Charts est interopérable avec n'importe quel organisation d'émission de charts pour peu qu'un connecteur existe. Un tel connecteur doit implémenter l'interface ChartsConnector et être enregistré comme un plugin du groupe charts_connector.

Les ChartsConnector doivent toujours fournir des charts avec une date d'expiration. Grâce à cela, charts peut conserver et ré-utiliser en local les charts récupérer depuis Internet. L'utilisation de la persistance est motivé par le fait que les charts sont releasé au maximum une fois par semaine. La persistance se fait à l'aide de 3 tables:
 - ChartsProvider représente une source de donnée. Par exemple, les charts quotidien pour la radio aux USA fourni par  WorldCharts.example.com constitue un ChartsProvider. WorldCharts.example.com seul n'est pas un ChartsProvider.
 - ChartsRelease représente la liste des entrées d'un charts pour un ChartsProvider. Une ChartsRelease doit disposer d'une date d'expiration.
 - ChartsEntry représente une entrée le classement. On y trouve les identifiants de l'œuvre (titre-artiste) et sa position dans le classement. On y trouve aussi sa meilleure position au classement depuis sa sortie et le nombre de semaine présent dans le classement. Je ne sais pas si on devrait laisser ces deux dernières informations.
 
 #### Le ChartsConnector à destination de la France
 SnepChartsConnector est un ChartsConnector qui va récupérer les charts officiel en France sur snepmusique.com via une simple extraction HTML.

**Comment by stevezau** [comment]
@gazpachoking you had a chance to review this yet?

**Comment by liiight** [comment]
@lildadou I'm starting to go over this beast :smile: 
really wanna merge this, this would be huge for flexget

File: flexget/plugins/filter/charts.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+@db_schema.upgrade('charts_snep')
+def upgrade(ver, session):
+    return 0
+
+
+class ChartsProvider(Base):
+    __tablename__ = 'charts_provider'
+    id = Column(Integer, primary_key=True)
+    organization = Column(String, nullable=False) # Company, association, person etc who provides charts
```
Comment (liiight) [inline flexget/plugins/filter/charts.py]: Shouldn't this be Unicode instead of String?

File: flexget/plugins/filter/charts.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        and the provider doesn't exist in DB then this method will create one and
+         return its id.
+        :type organization: str
+        :type category: str
+        :type periodicity: str
+        :type auto_create: bool
+        :rtype: int
+        """
+
+        session = Session()
```
Comment (liiight) [inline flexget/plugins/filter/charts.py]: I think it's preferable to use session as a context manager, more readable IMO:
```python
with Session as session:
```
That way you could ditch the try, finally method which works fine, but kinda hacky for this purpose.
Also you can use the `@with_session` decorator

File: flexget/plugins/filter/charts.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            log.debug("Attempting to retrieve charts online because %s" % online_reason)
+            last_release = self.get_last_release_online(organization, category, periodicity)
+            last_release.provider_id = provider_id
+            session.add(last_release)
+            session.commit()
+
+        return last_release
+
+    def on_task_filter(self, task, config):
+        session = Session()
```
Comment (liiight) [inline flexget/plugins/filter/charts.py]: please use `task.session` here

**Comment by liiight** [comment]
I seems you dropped a few unrelated plugins in the process. not sure why (sonnar_emit, letterboxed, etc.).

Also, I don't know about adding a separate file for audio qualities.
Maybe we should add a `type` attribute to quality component and have them belong to one or more types. Because what we currently would refer to as a video type quality component does share some audio components that could be used for a QualityRequirement of a music object. 
What do you think?

**Comment by lildadou** [comment]
Welcome on board. I will update the branch and take into account your code review. I think we will have to look at the title of parser guessit because the author has made major changes. (It is possible to bring this part of the code?)

See you at the exit of the tunnel :)

**Comment by lildadou** [comment]
@liiight : I do not know why I have deleted files but it's obviously a mistake.

**Comment by lildadou** [comment]
@liiight As I thought, this branch is too far back and you can not test yet. 

GuessIt has completely changed in the meantime. According to GuessIt author, I will integrate music parsing units directly in GuessIt. Then we can sync the branch.

**Comment by Toilal** [comment]
:+1:

**Comment by liiight** [comment]
ok, looking forward to it :smile:

**Comment by stevezau** [comment]
@lildadou just so you know we just moved from nose to pytest and are in the middle of enabling py3 support.. So when this merges in we need to keep that in mind..

**Comment by stevezau** [comment]
any progress here?

**Comment by lildadou** [comment]
Very few but WIP.

**Comment by liiight** [comment]
Maybe this pr could be split. Just the music parser alone and qualities have merit ob their own, and other devs can build off that

**Comment by lildadou** [comment]
@liiight : Yep, it's the current goal (https://github.com/Flexget/Flexget/pull/533#issuecomment-181875259)
Should we close the PR for the moment?

**Comment by liiight** [comment]
I guess so

**Pull request closed without merging**